### PR TITLE
fee middleware: grpc queries

### PR DESCRIPTION
**Pull request by charleenfei**
## Description

closes: #479 

Note that this PR adjusts the protos to reduce redundancy in the API.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

**Review by AdityaSripal — APPROVED** [review]
utACK

File: modules/apps/29-fee/keeper/grpc_query.go
Comment (seantking) [inline modules/apps/29-fee/keeper/grpc_query.go:18]: nit: I know that we're using the var `q` for the transfer query functions also but I find this confusing. 

@colin-axner is there any reason for this? I feel like using `k` or `keeper` like the rest of the SDK modules is easier to understand.

**Review by seantking — APPROVED** [review]
Nice work :)

Reminder: let's wait until the #259 PR is merged in and then we can target the feature branch

File: modules/apps/29-fee/keeper/grpc_query.go
Comment (damiannolan) [inline modules/apps/29-fee/keeper/grpc_query.go:10]: Super small nit: import ordering should be grouped as std lib, external libs, ibc-go imports, [see here](https://github.com/cosmos/ibc-go/pull/337#discussion_r694058339)

File: modules/apps/29-fee/keeper/grpc_query.go
Comment (damiannolan) [inline modules/apps/29-fee/keeper/grpc_query.go:23]: Another small nit:
This could be reduced to a single line and might improve readability by chaining `.WithBlockHeight()` like so:

```go
ctx := sdk.UnwrapSDKContext(goCtx).WithBlockHeight(int64(req.QueryHeight))
```

File: modules/apps/29-fee/keeper/grpc_query.go
Comment (colin-axner) [reply modules/apps/29-fee/keeper/grpc_query.go:18]: I'm not sure the reason. `k` sounds good

File: modules/apps/29-fee/keeper/grpc_query.go
Comment (colin-axner) [inline modules/apps/29-fee/keeper/grpc_query.go:22]:
```suggestion
	}
	
	ctx := sdk.UnwrapSDKContext(c)
```

File: modules/apps/29-fee/keeper/grpc_query.go
Comment (colin-axner) [inline modules/apps/29-fee/keeper/grpc_query.go:24]:
```suggestion
	ctx = ctx.WithBlockHeight(int64(req.QueryHeight))
	
	packets := []*types.IdentifiedPacketFee{}
```

File: modules/apps/29-fee/keeper/grpc_query.go
Comment (colin-axner) [inline modules/apps/29-fee/keeper/grpc_query.go:50]:
```suggestion
	ctx := sdk.UnwrapSDKContext(c)
	ctx = ctx.WithBlockHeight(int64(req.QueryHeight))
	
	fee, exists := q.GetFeeInEscrow(ctx, req.PacketId)
```

File: modules/apps/29-fee/keeper/grpc_query.go
Comment (colin-axner) [inline modules/apps/29-fee/keeper/grpc_query.go:48]:
```suggestion
	}
	
	ctx := sdk.UnwrapSDKContext(c)
```

File: modules/apps/29-fee/keeper/grpc_query_test.go
Comment (colin-axner) [inline modules/apps/29-fee/keeper/grpc_query_test.go:18]: These should be kept within scope of the test function they are used in (otherwise leftover assignments could unintentionally be used between tests)

File: modules/apps/29-fee/keeper/grpc_query_test.go
Comment (colin-axner) [inline modules/apps/29-fee/keeper/grpc_query_test.go:29]: nit: We should add a `NewPacketID` function and use that instead

File: modules/apps/29-fee/keeper/grpc_query_test.go
Comment (colin-axner) [inline modules/apps/29-fee/keeper/grpc_query_test.go:27]: nit: use ibctesting.FirstChannelID (no need to create new variable)

File: modules/apps/29-fee/keeper/grpc_query_test.go
Comment (colin-axner) [inline modules/apps/29-fee/keeper/grpc_query_test.go:38]:
```suggestion
	coins := sdk.Coins{sdk.Coin{Denom: sdk.DefaultBondDenom, Amount: sdk.NewInt(100)}}
	fee := &types.Fee{
		AckFee:     coins,
		ReceiveFee: coins,
		TimeoutFee: coins,
	}
```

File: modules/apps/29-fee/keeper/grpc_query_test.go
Comment (colin-axner) [inline modules/apps/29-fee/keeper/grpc_query_test.go:124]: nit: use NewIdentifiedPacketFee

**Review by colin-axner — COMMENTED** [review]
Nice work! left several nits/suggestions. Feel free to pick and choose which to apply as you please

File: modules/apps/29-fee/keeper/keeper_test.go
Comment (seantking) [inline modules/apps/29-fee/keeper/keeper_test.go:71]: This function should go in the `types` package similar to other helpers. see -> https://github.com/cosmos/ibc-go/blob/ics29-fee-middleware/modules/apps/29-fee/types/msgs.go#L17

File: modules/apps/29-fee/keeper/keeper_test.go
Comment (seantking) [inline modules/apps/29-fee/keeper/keeper_test.go:66]: This should actually go into main and update the channel types here probably -> https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/types/packet.go 

I think you can leave this function out for now and I'll make an issue for it and tag @colin-axner

**Review by seantking — CHANGES_REQUESTED** [review]
Just need to move the `NewIdentifiedPacketFee` function into the types package.

File: modules/apps/29-fee/keeper/keeper_test.go
Comment (charleenfei) [reply modules/apps/29-fee/keeper/keeper_test.go:71]: it's just a helper function for testing though, if we want to make it a full helper i would add the relayers array into the function params

File: modules/apps/29-fee/keeper/keeper_test.go
Comment (charleenfei) [reply modules/apps/29-fee/keeper/keeper_test.go:66]: same comment as above but with channel-id

File: modules/apps/29-fee/keeper/grpc_query.go
Comment (damiannolan) [inline modules/apps/29-fee/keeper/grpc_query.go:27]: This is a super subtle difference, but follows along with other code in the repo and is probably considered more "standard". Essentially its declaration vs initialisation.

```suggestion
	var packets []*types.IdentifiedPacketFee
```

File: modules/apps/29-fee/keeper/grpc_query_test.go
Comment (damiannolan) [inline modules/apps/29-fee/keeper/grpc_query_test.go:123]: `append()` takes a variadic number of args so you can pass all three fees one after the other:

```
expPackets = append(expPakets, &fee1, &fee2, &fee3)
```

File: modules/apps/29-fee/keeper/grpc_query_test.go
Comment (damiannolan) [inline modules/apps/29-fee/keeper/grpc_query_test.go:115]: You can call `NewPacketId` inside of `NewIdentifiedPacketFee` to avoid redundant variable assignments here too.
You should also be able to drop the `uint64()` conversion afaik, as the arg is defined as `uint64`

```
NewIdentifiedPacketFee(NewPacketId(1), *fee)
```

File: modules/apps/29-fee/keeper/keeper_test.go
Comment (damiannolan) [reply modules/apps/29-fee/keeper/keeper_test.go:71]: I think this should exist in `package types`, it's wrong for it to exist as a method on the KeeperTestSuite. 

Right now it's just a helper function for testing, but I think in your follow-up PR for calling the escrow logic funcs you could make use of it and then it wouldn't make sense to have it here - e.g. [`EscrowPacketFee`](https://github.com/cosmos/ibc-go/blob/ics29-fee-middleware/modules/apps/29-fee/keeper/escrow.go#L15) you may want to use `NewIdentifiedPacketFee()` to construct that type, and pass it as an arg

**Review by damiannolan — COMMENTED** [review]
Left a couple of more small nits, happy to approve when @seantking's concern is addressed. 
Nice work!

File: modules/apps/29-fee/keeper/keeper_test.go
Comment (seantking) [reply modules/apps/29-fee/keeper/keeper_test.go:66]: Having it in the testing file is just duplicate code. We should have these in types similar to all of our other proto helpers. 

You can do this in a separate PR though. Agree with @damiannolan  :+1:

**Review by seantking — APPROVED** [review]

File: modules/apps/29-fee/keeper/keeper_test.go
Comment (seantking) [reply modules/apps/29-fee/keeper/keeper_test.go:71]: https://github.com/cosmos/ibc-go/blob/issue%23479-grpc-query/modules/apps/29-fee/types/msgs.go#L165

I actually already added this on the previous PR. You can use this. Sorry, I just realized :)

**Pull request merged**
